File: cpp/se/EnumDomainCreateCommand.cpp
#include "se/EnumDomainCreateCommand.hpp"
#include "se/E164Extension.hpp"

namespace {
    Extension& e164Extension() {
        static Extension* extension = new E164Extension();
        return *extension;
    }
}; // anonymous namespace

EnumDomainCreateCommand::EnumDomainCreateCommand (
        const std::string &name,
        const std::string &pw,
        const std::string *registrantID,
        const std::vector<std::string> &techContacts) : DomainCreateCommand (
            name, pw, registrantID, &techContacts)
{ }


EnumDomainCreateCommand::EnumDomainCreateCommand (
        const std::string &name,
        const std::string &pw,
        const std::string *registrantID,
        const std::vector<std::string> &techContacts,
        const std::vector<NAPTR> *naptrs) : DomainCreateCommand (
            name, pw, registrantID, &techContacts)
{
    setExtension (naptrs);
}

EnumDomainCreateCommand::EnumDomainCreateCommand (
        const std::string &name,
        const std::string &pw,
        const std::string *registrantID,
        const std::vector<std::string> &techContacts,
        const std::vector<std::string> &adminContacts,
        const std::vector<std::string> &billingContacts,
        const std::vector<NAPTR> *naptrs,
        const Period &period) : DomainCreateCommand (
            name, pw, registrantID, &techContacts,
            &adminContacts, &billingContacts,
            NULL, &period)
{
    setExtension (naptrs);
}     

EnumDomainCreateCommand::EnumDomainCreateCommand (
        const std::string &name,
        const std::string &pw,
        const std::string *registrantID,
        const std::vector<std::string> &techContacts,
        const std::vector<std::string> &adminContacts,
        const std::vector<std::string> &billingContacts,
        const std::vector<std::string> &nameservers,
        const Period &period) : DomainCreateCommand (
            name, pw, registrantID, &techContacts,
            &adminContacts, &billingContacts, &nameservers,
            &period)
{
}

void EnumDomainCreateCommand::setExtension (const std::vector<NAPTR> *naptrs)
{
    if (naptrs == NULL || naptrs->size() == 0)
        return;

    DOMElement *e164Create = xmlWriter->appendChild(
            xmlWriter->appendChild(
                command,
                "extension"),
            "create",
            e164Extension().getURI());

    std::vector<NAPTR>::const_iterator p = naptrs->begin();

    while (p != naptrs->end())
    {
        p->appendToElement (xmlWriter, e164Create);
        p++;
    }
}

File: src/main/java/com/ausregistry/jtoolkit2/se/EnumDomainCreateCommand.java
package com.ausregistry.jtoolkit2.se;

import org.w3c.dom.Element;


/**
 * Use this to request provisioning of an ENUM domain object in an EPP
 * Registry.  Instances of this class generate domain create EPP service
 * elements compliant with RFC5730, RFC5731 and RFC4114 via the toXML method.
 */
public final class EnumDomainCreateCommand extends DomainCreateCommand {
    private static final long serialVersionUID = 1661618997961186028L;

    /**
     * Minimal constructor for creating a domain:create + e164epp:create
     * EPP command.  These parameters are the least required for a valid
     * ENUM domain create command.
     */
    public EnumDomainCreateCommand(String name, String pw,
            String registrantID, String[] techContacts) {
        super(name, pw, registrantID, techContacts);
    }

    /**
     * Construct a domain:create + e164epp:create EPP command with NAPTR
     * records.  This is the least information required to provision an ENUM
     * domain with NAPTR records.
     */
    public EnumDomainCreateCommand(String name, String pw,
            String registrantID, String[] techContacts,
            NAPTR[] naptrs) {
        super(name, pw, registrantID, techContacts);
        setExtension(naptrs);
    }

    /**
     * Full data specification constructor for a domain:create + e164epp:create
     * EPP command with NAPTR records.  Please refer to the
     * urn:ietf:params:xml:ns:e164epp-1.0 schema for specification of the
     * required fields.
     */
    public EnumDomainCreateCommand(String name, String pw, String registrantID,
            String[] techContacts, String[] adminContacts,
            String[] billingContacts,
            NAPTR[] naptrs, Period period) {
        super(name, pw, registrantID,
                techContacts, adminContacts, billingContacts,
                EMPTY_NAMESERVERS, period);
        setExtension(naptrs);
    }

    /**
     * Full data specification constructor for a domain:create + e164epp:create
     * EPP command with nameservers rather than NAPTR records.  This
     * constructor does not cause the e164epp extension element to be created,
     * since NAPTR records are not specified.
     */
    public EnumDomainCreateCommand(String name, String pw, String registrantID,
            String[] techContacts, String[] adminContacts,
            String[] billingContacts,
            String[] nameservers, Period period) {
        super(name, pw, registrantID, techContacts,
                adminContacts, billingContacts, nameservers, period);
    }

    private void setExtension(NAPTR[] naptrs) {
        if (naptrs == null || naptrs.length == 0) {
            return;
        }

        extension = xmlWriter.appendChild(command, "extension");
        Element extensionElement = xmlWriter.appendChild(extension, "create", ExtensionImpl.E164.getURI());

        for (NAPTR naptr : naptrs) {
            naptr.appendToElement(xmlWriter, extensionElement);
        }
    }
}

